New Noisy Force Langevin Thermostat

Ithanil opened the pull request:
Added a new thermostat named ThermoNFL (Noisy Force Langevin). It is a modified Langevin thermostat, which compensates for forces with inherent white noise by adequate additional damping. Includes a method to automatically adjust the required force noise sigma parameter.

Using the labels T for damping factor and S for noise coefficient (S_i for inherent white noise in forces, S_a for additional white noise from thermostat) like in the code, it can be described as follows:
On top of the inherent white noise of the forces with coefficient S_i, the thermostat adds additional white noise with coefficient S_a. This means we have to damp according to a combined white noise with coefficient S_c^2 = S_a^2 + S_i^2. The damping T is related to S_c as S_c^2 = kb * temp * (1 - T^2) <=> T^2 = 1 - S_c^2 / kb / temp = 1 - (S_a^2 + S_i^2)/ kb / temp = exp(- dt / tau) - S_i^2 / kb / temp . 
The S_i is called "sigma" in the code, is an input parameter and has to be adjusted to the amplitude of force noise. Alternatively you can additionally provide a "sigtau" parameter to have it adjusted automatically in every step as follows:
sigma += (ctemp - ttemp) / sigtau * dt , 
where ctemp is the calculated temperature of the system in the current timestep and temp is the target temperature from input. This part is executed only once per timestep. At the moment the current sigma is simply printed out to console to check for it.
 
So far the thermostat works as expected, but was only used by myself. It's the first pull request I ever wrote so I hope it is done adequately. Feel free to suggest improvements!

ceriottm commented on ipi/engine/thermostats.py:
```diff
@@ -194,7 +194,104 @@ def step(self):
+   def get_T(self):
+      """Calculates the coefficient of the overall drift of the velocities."""
+
+      SigT = self.sigma**2/(Constants.kb*self.temp)
```
Are you sure about this line? This does not make dimensional sense, T is a pure number, sigma is (according to the input definition) a force, and kT is an energy.

ceriottm commented on ipi/engine/thermostats.py:
```diff
@@ -194,7 +194,104 @@ def step(self):
+      if self.sigtau > 0 and self.sigstep:
+         ekin = np.dot(depstrip(self.p),depstrip(self.p)/depstrip(self.m))*0.5
+         mytemp = ekin/Constants.kb/self.ndof * 2
+         self.sigma += (mytemp - self.temp) / self.sigtau * self.dt
```
Also here, sigma should be a force, and you say that it equals a temperature divided time square.

Ithanil commented:
Thanks a lot for the feedback, although I should have checked the dimensional consistency already by myself. Sorry for that!
As you pointed out, the old sigma**2 must be an energy. So to fix consistency I replaced the 'sigma' variable by the new 'invar' (inherent noise variance), where invar =  sigma**2 . Changed the corresponding dimension declaration in input/thermostats.py accordingly.
Also the old 'sigtau' adjustment time coefficient is replaced by 'invtau'. Finally the 'invar' adjustment step now contains a Boltzmann factor to fix the dimensional issue there.
Changes are working as expected for my testing toy system.

Ithanil commented:
I'm still not able to run my previously working input files since including the commits concerning binary I/O backend. i-pi seems to hang after handshaking and assigning client ID. This is also happening using unmodified clones from upstream master, so it seems it is not related to my own changes. Should I open a separate issue or is it already known that the current version has problems?

Here is a paste of the output:
`[ithanil@arbeitsmilch testdir_dev]$ python2 ~/Downloads/i-pi-dev/bin/i-pi lgvtest.xml

 ____       ____       ____       ____
/    \     /    \     /    \     /    \ 
|  #################################  |
\__#_/     \____/     \____/     \_#__/
   #    _        _______  _____    #
   #   (_)      |_   __ \|_   _|   #      -*-     Development version    -*-
   #   __  ______ | |__) | | |     #
   Y  [  ||______||  ___/  | |     #      A Python interface for (ab initio)
  0 0  | |       _| |_    _| |_    #      (path integral) molecular dynamics.
   #  [___]     |_____|  |_____|   #
 __#_       ____       ____       _#__
/  # \     /    \     /    \     / #  \
|  #################################  |
\____/     \____/     \____/     \____/

    
 # Initializing system object 
 # Initializing simulation object 
 # Initializer (stage 1) parsing positions object.
 # Initializer (stage 1) parsing masses object.
 # Initializer (stage 1) parsing labels object.
 # Initializer (stage 1) parsing cell object.
 # Initializer (stage 1) parsing velocities object.
 # Resampling velocities at temperature 1000.0 kelvin
Import of PyFFTW unsuccessful, using NumPy library instead
 # Initializer (stage 1) parsing cell object.
 !W! Overwriting previous cell parameters
Import of PyFFTW unsuccessful, using NumPy library instead
 # Initializer (stage 2) parsing positions object.
 # Initializer (stage 2) parsing masses object.
 # Initializer (stage 2) parsing labels object.
 # Initializer (stage 2) parsing cell object.
 # Initializer (stage 2) parsing velocities object.
 # Initializer (stage 2) parsing cell object.
Backup performed: sampletest.md -> #sampletest.md#13#
 # i-PI loaded input file:  lgvtest.xml
 --- begin input file content ---
<simulation verbosity='high'>
   <output prefix='sampletest'>
      <properties filename='md' stride='2'> [step, time{femtosecond}, temperature{kelvin}, potential{atomic_unit}, kinetic_md{atomic_unit}] </properties>
   </output>
   <total_steps> 200 </total_steps>
   <ffsocket mode="inet" name="driver">
      <address>localhost</address>
      <port> 11111 </port>
   </ffsocket>
   <system>
      <initialize nbeads='1'>
        <file mode='xyz'> start0.xyz </file>
        <velocities mode='thermal' units='kelvin'> 1000 </velocities>
        <cell mode='abc' units='atomic_unit'> [1, 1, 1] </cell> 
      </initialize>
      <forces><force forcefield='driver'></force></forces>
      <ensemble>
         <temperature units='kelvin'> 1000 </temperature>
      </ensemble>
      <motion mode='dynamics'>
	<dynamics mode='nvt'>
	  <thermostat mode='langevin'>
            <tau units='femtosecond'> 0.1 </tau>
          </thermostat>
          <timestep units='femtosecond'> 0.5 </timestep>
	</dynamics>
      </motion>
   </system>
</simulation>
 ---  end input file content  ---
Created inet socket with address localhost and port number 11111
 @ForceField: Starting the polling thread main loop.
 @SOCKET:   Client asked for connection from ('127.0.0.1', 39020). Now hand-shaking.
 @SOCKET:   Handshaking was successful. Added to the client list.
 @SOCKET: 17/05/31-17:05:30 Assigning [ none] request id    0 to client with last-id None (  0/  1 : ('127.0.0.1', 39020))`

ceriottm commented:
Hi Jan, I'm sorry but all the developers are very busy these days, it would
be nice if you could try to track this down a bit further as it will take
some time before any of us can look into this problem. Cheers

On 31 May 2017 at 17:29, Jan Kessler <notifications@github.com> wrote:

> I'm still not able to run my previously working input files since
> including the commits concerning binary I/O backend. i-pi seems to hang
> after handshaking and assigning client ID. This is also happening using
> unmodified clones from upstream master, so it seems it is not related to my
> own changes. Should I open a separate issue or is it already known that the
> current version has problems?
>
> Here is a paste of the output:
> `[ithanil@arbeitsmilch testdir_dev]$ python2
> ~/Downloads/i-pi-dev/bin/i-pi lgvtest.xml
> ------------------------------
>
> / \ / \ / \ / \
> | ################################# |
> _*#*/ ____/ __*/ _#*/
> _ _______ _____ (*) |* __ |_ _| # -*- Development version -*- __ ______ |
> |__) | | |
>
> Y [ ||______|| ___/ | | # A Python interface for (ab initio)
> 0 0 | | *| |* *| |* # (path integral) molecular dynamics.
> [*] |**| |*___|
>
> *# ____ ____ #*
> / # \ / \ / \ / #
> | ################################# |
> ____/ ____/ ____/ ____/
> Initializing system object Initializing simulation object Initializer
> (stage 1) parsing positions object. Initializer (stage 1) parsing masses
> object. Initializer (stage 1) parsing labels object. Initializer (stage
> 1) parsing cell object. Initializer (stage 1) parsing velocities object. Resampling
> velocities at temperature 1000.0 kelvin
>
> Import of PyFFTW unsuccessful, using NumPy library instead
> Initializer (stage 1) parsing cell object.
>
> !W! Overwriting previous cell parameters
> Import of PyFFTW unsuccessful, using NumPy library instead
> Initializer (stage 2) parsing positions object. Initializer (stage 2)
> parsing masses object. Initializer (stage 2) parsing labels object. Initializer
> (stage 2) parsing cell object. Initializer (stage 2) parsing velocities
> object. Initializer (stage 2) parsing cell object.
>
> Backup performed: sampletest.md -> #sampletest.md#13#
> <http://sampletest.md#13%23>
> i-PI loaded input file: lgvtest.xml
>
> --- begin input file content ---
>
>
> [step, time{femtosecond}, temperature{kelvin}, potential{atomic_unit},
> kinetic_md{atomic_unit}]
>
> <total_steps> 200 </total_steps>
>
> localhost
> 11111
>
>
>
> start0.xyz
> 1000
> [1, 1, 1]
>
>
>
> 1000
>
>
>
>
> 0.1
>
> 0.5
>
>
>
>
> --- end input file content ---
> Created inet socket with address localhost and port number 11111
> @forcefield <https://github.com/forcefield>: Starting the polling thread
> main loop.
> @socket <https://github.com/socket>: Client asked for connection from
> ('127.0.0.1', 39020). Now hand-shaking.
> @socket <https://github.com/socket>: Handshaking was successful. Added to
> the client list.
> @socket <https://github.com/socket>: 17/05/31-17:05:30 Assigning [ none]
> request id 0 to client with last-id None ( 0/ 1 : ('127.0.0.1', 39020))`
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/cosmo-epfl/i-pi-dev/pull/184#issuecomment-305224369>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABESZ04XewbzAxO2bLWKsYEq7deL3hU6ks5r_Yd1gaJpZM4NSDTZ>
> .
>

Ithanil commented:
Hello Michele,
good that noone wasted time on this... Problem was related to my testing driver and a commit that is included in my qmc-pimd branch but intentionally not in my master branch. But because I noticed it coincidentally right after merging upstream, I was fooled into suspecting the upstream changes to cause the problems.
So: Actually works as intended.

venkatkapil24 commented:
Hey @Ithanil, It would be great if you could provide us an example  that works with cp2k. You can place it in examples/cp2k

Ithanil commented:
Hello @venkatkapil24 , thank you for looking into this. But unfortunately providing a cp2k example is not straightforward for me, because I do not and never did use cp2k. Instead I built this thermostat to compensate for the stochastic nature of forces from QMC calculations.
However, for testing purposes I built a toy system driver, which calculates forces for a system of N uncoupled 3D harmonic oscillators and adds noise to those forces before sending them to i-pi. If you are interested, I could create an example out of that.

ceriottm commented:
Even better, if the QMC code is public and the i-PI interface is officially
part of the distribution, you could add an example for this QMC code.
M

On 19 June 2017 at 16:13, Jan Kessler <notifications@github.com> wrote:

> Hello @venkatkapil24 <https://github.com/venkatkapil24> , thank you for
> looking into this. But unfortunately providing a cp2k example is not
> straightforward for me, because I do not and never did use cp2k. Instead I
> built this thermostat to compensate for the stochastic nature of forces
> from QMC calculations.
> However, for testing purposes I built a toy system driver, which
> calculates forces for a system of N uncoupled 3D harmonic oscillators and
> adds noise to those forces before sending them to i-pi. If you are
> interested, I could create an example out of that.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/cosmo-epfl/i-pi-dev/pull/184#issuecomment-309452251>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABESZxQkvopQtUcs8hNBV-yPYhqwjTrEks5sFoIHgaJpZM4NSDTZ>
> .
>

Ithanil commented:
I am working on Francesco Calcavecchia's [https://github.com/francesco086/HswfQMC](url) , but the force functionality and i-pi interface is added by me, not yet in master and still WIP. That's why I thought the toy system example might be better for now.

ceriottm commented:
It would be excellent if you could accelerate the deployment of the force
interface in the mainline code, as that would be a nice addition.
Generally speaking, with the new release upcoming, I am trying to get as
many concrete contributions included, as that provides a citation point for
future work (i.e. you could ask people using the interface in the future to
give you a cite for that, regardless of whether they use the NFLT part).
Let me know if you think this is feasible.
Best
M

On 19 June 2017 at 16:49, Jan Kessler <notifications@github.com> wrote:

> I am working on Francesco Calcavecchia's https://github.com/
> francesco086/HswfQMC <http://url> , but the force functionality and i-pi
> interface is added by me, not yet in master and still WIP. That's why I
> thought the toy system example might be better for now.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/cosmo-epfl/i-pi-dev/pull/184#issuecomment-309463339>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABESZ7qlXqAnGEch9TLxx6xG4ZNNHFMoks5sFoppgaJpZM4NSDTZ>
> .
>

Ithanil commented:
The i-pi interface for HswfQMC is in principle ready to be deployed to master. However, the gradients for the forces are not fully exact yet and we are still working on that part. Also some general polish is missing. Or stated differently, the whole package is not exactly ready for general use yet, but hopefully we will be there soon. Best wishes, Jan

Ithanil commented:
Or to make it more clear: I will deploy the interface functionality to master very soon, but the gradients when we are sure about them.

Ithanil commented:
Finally I got some news regarding the proposed HswfQMC example:

Because the original author, Francesco Calcavecchia, is currently not available to review a master-merge of all the changes to HswfQMC, I resorted to creating a 'release' branch which acts as a sort of release candidate and contains the functionality required for an HswfQMC/i-pi example. New features will be developed outside of that branch and it is intended to be merged to master as soon as possible. 

Based on the release version of HswfQMC I created a simple example of a HswfQMC/i-pi MD with enabled NFL Thermostat. It can be found in examples/hswfqmc together with a README.
For the sake of balance between computational effort and usability of the variance parameter adjustment algorithm within the NFL thermostat, the example concerns a rather questionable system of 8 hydrogen atoms at temperature 5000K and r_s=1.31. Also, as said before, the accuracy of the forces itself is still not fully established. Concluding, the example is meant to be a demonstration of HswfQMC's i-pi interface and the new NFL thermostat's functionality and not to provide a template for actual production runs as it is.

The provided example demonstrates only classical MD with fixed wavefunction. PI functionality is currently disabled in the release client, because it relies on an additional communication of the beadid between client and server in every step. Doing it that way obviously breaks all other clients. When I found a better way of telling the client the current beadid and fixed it, I could provide another example with 4 bead PI or something like that.

So, now I hope that the provided instructions allow you to run the example as intended and in the end a merge of my fork can take place.

Best wishes,
Jan

ceriottm commented:
Actually you can get beadid in the "INITIALIZE" field. I think the quantum
Espresso module uses this to decide whether to extrapolate or reset. Have a
look at that before merging a version that is not fully functional.
Cheers
Michele

On 24 July 2017 at 09:48, Jan Kessler <notifications@github.com> wrote:

> Finally I got some news regarding the proposed HswfQMC example:
>
> Because the original author, Francesco Calcavecchia, is currently not
> available to review a master-merge of all the changes to HswfQMC, I
> resorted to creating a 'release' branch which acts as a sort of release
> candidate and contains the functionality required for an HswfQMC/i-pi
> example. New features will be developed outside of that branch and it is
> intended to be merged to master as soon as possible.
>
> Based on the release version of HswfQMC I created a simple example of a
> HswfQMC/i-pi MD with enabled NFL Thermostat. It can be found in
> examples/hswfqmc together with a README.
> For the sake of balance between computational effort and usability of the
> variance parameter adjustment algorithm within the NFL thermostat, the
> example concerns a rather questionable system of 8 hydrogen atoms at
> temperature 5000K and r_s=1.31. Also, as said before, the accuracy of the
> forces itself is still not fully established. Concluding, the example is
> meant to be a demonstration of HswfQMC's i-pi interface and the new NFL
> thermostat's functionality and not to provide a template for actual
> production runs as it is.
>
> The provided example demonstrates only classical MD with fixed
> wavefunction. PI functionality is currently disabled in the release client,
> because it relies on an additional communication of the beadid between
> client and server in every step. Doing it that way obviously breaks all
> other clients. When I found a better way of telling the client the current
> beadid and fixed it, I could provide another example with 4 bead PI or
> something like that.
>
> So, now I hope that the provided instructions allow you to run the example
> as intended and in the end a merge of my fork can take place.
>
> Best wishes,
> Jan
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/cosmo-epfl/i-pi-dev/pull/184#issuecomment-317466036>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABESZ_yIobhivYkoOyIJJ-whmXprWUtDks5sRLzfgaJpZM4NSDTZ>
> .
>

mahrossi commented:
We also use this functionality in FHI-aims! (getting the bead ID)

On Mon, Jul 24, 2017 at 6:00 PM, Michele Ceriotti <notifications@github.com>
wrote:

> Actually you can get beadid in the "INITIALIZE" field. I think the quantum
> Espresso module uses this to decide whether to extrapolate or reset. Have a
> look at that before merging a version that is not fully functional.
> Cheers
> Michele
>
>
> On 24 July 2017 at 09:48, Jan Kessler <notifications@github.com> wrote:
>
> > Finally I got some news regarding the proposed HswfQMC example:
> >
> > Because the original author, Francesco Calcavecchia, is currently not
> > available to review a master-merge of all the changes to HswfQMC, I
> > resorted to creating a 'release' branch which acts as a sort of release
> > candidate and contains the functionality required for an HswfQMC/i-pi
> > example. New features will be developed outside of that branch and it is
> > intended to be merged to master as soon as possible.
> >
> > Based on the release version of HswfQMC I created a simple example of a
> > HswfQMC/i-pi MD with enabled NFL Thermostat. It can be found in
> > examples/hswfqmc together with a README.
> > For the sake of balance between computational effort and usability of the
> > variance parameter adjustment algorithm within the NFL thermostat, the
> > example concerns a rather questionable system of 8 hydrogen atoms at
> > temperature 5000K and r_s=1.31. Also, as said before, the accuracy of the
> > forces itself is still not fully established. Concluding, the example is
> > meant to be a demonstration of HswfQMC's i-pi interface and the new NFL
> > thermostat's functionality and not to provide a template for actual
> > production runs as it is.
> >
> > The provided example demonstrates only classical MD with fixed
> > wavefunction. PI functionality is currently disabled in the release
> client,
> > because it relies on an additional communication of the beadid between
> > client and server in every step. Doing it that way obviously breaks all
> > other clients. When I found a better way of telling the client the
> current
> > beadid and fixed it, I could provide another example with 4 bead PI or
> > something like that.
> >
> > So, now I hope that the provided instructions allow you to run the
> example
> > as intended and in the end a merge of my fork can take place.
> >
> > Best wishes,
> > Jan
> >
> > —
> > You are receiving this because you commented.
> > Reply to this email directly, view it on GitHub
> > <https://github.com/cosmo-epfl/i-pi-dev/pull/184#issuecomment-317466036
> >,
> > or mute the thread
> > <https://github.com/notifications/unsubscribe-
> auth/ABESZ_yIobhivYkoOyIJJ-whmXprWUtDks5sRLzfgaJpZM4NSDTZ>
> > .
>
> >
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/cosmo-epfl/i-pi-dev/pull/184#issuecomment-317469526>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AGbkQn7SeNnRWJKYhjGEI1yxmoe2f1OXks5sRL-HgaJpZM4NSDTZ>
> .
>

Ithanil commented:
Thanks for the quick response. Actually my hope was to somehow get that information via NEEDINIT / INIT . Now I know how easy it actually is :) However I won't get it done today, but certainly push an update tomorrow.
Thanks again,
Jan

Ithanil commented:
But remember that I'm using an adaption of an old version of your driver code as part of HswfQMC and didn't add HswfQMC as an option to i-pi's own driver. So the update to my i-pi fork will only contain another example with nbeads > 1. The code change takes place in HswfQMC's release branch.

ceriottm commented:
Sure. Just want to make sure that if you need changes here they are made
before merging. If you think you'll only need to add a second example, then
just add the example here and then it will become functional once you patch
HswfQMC


On 24 July 2017 at 10:30, Jan Kessler <notifications@github.com> wrote:

> But remember that I'm using an adaption of an old version of your driver
> code as part of HswfQMC and didn't add HswfQMC as an option to i-pi's own
> driver. So the update to my i-pi fork will only contain another example
> with nbeads > 1. The code change takes place in HswfQMC's release branch.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/cosmo-epfl/i-pi-dev/pull/184#issuecomment-317478282>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABESZwieKsyJI6nS-s0fLc5QF70u_slcks5sRMacgaJpZM4NSDTZ>
> .
>

Ithanil commented:
Second example with 2-bead PI and enabled simultaneous wavefunction optimization is added. HswfQMC's ipi client in release branch is changed to make it work. 
But I was surprised by an unexpected problem: 
It seems like the i-pi server wants to hand out INIT data only once per client. Sending NEEDINIT a second time from the same client resulted in hanging. Obviously in the "official" i-pi driver code NEEDINIT is sent once at max. But can I be sure that the bead<->client assignment will never change?
Whatever the answer is, concerning this merge I think I've got everything ready.
Cheers,
Jan

ceriottm commented:
it should be possible to send multiple NEEDINIT. as Mariana said, FHI-AIMS
does it
you're probably missing some bits in the communication protocol.

On 25 July 2017 at 10:05, Jan Kessler <notifications@github.com> wrote:

> Second example with 2-bead PI and enabled simultaneous wavefunction
> optimization is added. HswfQMC's ipi client in release branch is changed to
> make it work.
> But I was surprised by an unexpected problem:
> It seems like the i-pi server wants to hand out INIT data only once per
> client. Sending NEEDINIT a second time from the same client resulted in
> hanging. Obviously in the "official" i-pi driver code NEEDINIT is sent once
> at max. But can I be sure that the bead<->client assignment will never
> change?
> Whatever the answer is, concerning this merge I think I've got everything
> ready.
> Cheers,
> Jan
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/cosmo-epfl/i-pi-dev/pull/184#issuecomment-317784229>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABESZzdOGpxqU6qn1HAbvXT4b-DUGBvOks5sRhJggaJpZM4NSDTZ>
> .
>

Ithanil commented:
Yeah you are right. Somehow I missed that setting isinit=.FALSE. already during force calculation prevents the client from actually sending the forces, because isinit is checked before hasdata -.-
Works now as I wanted it, thanks :)

ceriottm commented:
OK. I guess this can be merged then. If you create an entry discussing briefly the method, following this template http://ipi-code.org/assets/pdf/example_feature.md I'll add a mention to it in the features page http://ipi-code.org/about/features/

Ithanil commented:
After talking back to Thomas Kühne I finished the following entry:
[noisy_force_thermostat.txt](https://github.com/cosmo-epfl/i-pi-dev/files/1183616/noisy_force_thermostat.txt) ( had to change from .md to .txt to upload oO )
There exists no paper with respect to this implementation (yet), but when that changes we would of course like to add an "Implementation" section.
Thanks for everything and best wishes,
Jan
